ancv/web/server.py

```diff
@@ -261,3 +264,124 @@ def server_timing_header(timings: dict[str, timedelta]) -> str:
         f"{name.replace(' ', '-')};dur={duration // timedelta(milliseconds=1)}"
         for name, duration in timings.items()
     )
+
+
+class WebHandler(Runnable):
+    """A handler serving a rendered template loaded from a URL with periodic refresh."""
+
+    def __init__(self, destination: str, refresh_interval: int = 300) -> None:
+        """Initializes the handler.
+
+        Args:
+            destination: The URL to load the JSON Resume from.
+            refresh_interval: How often to refresh the resume in seconds.
+        """
+        self.destination = destination
+        self.refresh_interval = refresh_interval
+        self.cache: Optional[str] = None
+        self.last_fetch: Optional[float] = None
+
+        LOGGER.debug("Instantiating web application.")
+        self.app = web.Application()
+
+        LOGGER.debug("Adding routes.")
+        self.app.add_routes([web.get("/", self.root)])
+
+        self.app.cleanup_ctx.append(self.app_context)
+
+    def run(self, context: ServerContext) -> None:
+        LOGGER.info("Loaded, starting server...")
+        web.run_app(self.app, host=context.host, port=context.port, path=context.path)
+
+    async def app_context(self, app: web.Application) -> AsyncGenerator[None, None]:
+        """Sets up the application context with required clients.
+
+        Args:
+            app: The app instance to attach our state to.
+        """
+        log = LOGGER.bind(app=app)
+        log.debug("App context initialization starting.")
+
+        log.debug("Starting client session.")
+        session = ClientSession()
+        app["client_session"] = session
+        log.debug("Started client session.")
+
+        log.debug("App context initialization done, yielding.")
+        yield
+
+        log.debug("App context teardown starting.")
+        await session.close()
+        log.debug("App context teardown done.")
+
+    async def fetch_and_render(self, session: ClientSession) -> str:
```

Comment:
It's probably a good idea to split `fetch` and `render` up. `render` is a pure function IIRC, and `fetch` will do I/O.

You can try and introduce a new top-level function which calls three separate ones: fetch, render, cache. Not totally sure that will be an improvement or even necessary, but splitting `fetch_and_render` is probably a good idea! We can experiment with this. It will make testing much simpler probably.

Reply:
I like the idea to split, makes much more sense and keep it cleaner and easier to maintain, done.